README.md
[![made-with-python](https://img.shields.io/badge/Made%20with-Python-1f425f.svg)](https://www.python.org/)
[![Build Status](https://travis-ci.org/datawrestler/sec-utils.svg?branch=master)](https://travis-ci.org/datawrestler/sec-utils)
[![PyPI version](https://badge.fury.io/py/secutils.svg)](https://badge.fury.io/py/secutils)

#### Welcome to [secutils](https://github.com/datawrestler/sec-utils)
**secutils** is a utility package to facilitate large bulk downloads of SEC documents. It works with any SEC document type and will retrieve the entire historical database if required. Multi-threaded file downloads are enabled in the command line utility.

Key functionality includes:
- Multi-threaded downloading
- Caching of index files
- Automatic directory structure buildout (i.e. downloading multiple file types w/dir structure: ftype --> year --> quarter --> files)
- Resume downloading
- Built in logging and download success tracker

Overview of README:
- [Motivation](#motivate)
- [Installation](#install)
- [Usage](#usage)
- [Vision](#vision)

##### Motivation <a id='motivate' />
secutils picks up where a number of other repos left off. There are a couple SEC downloading python packages out there, however they are designed from retrieval of few documents. I needed a way to consistently download the latest updates from the SEC and secure a local copy of the entire history of the SEC for certain file types. This translates into TB's of documents, where networking, directory structure, logging, etc. issues arise. 

There is a nice package available to download and construct index files, however the user is still left to download the actual files and must be comfortable with bash scripting. 

With secutils the program handles files you have already retrieved, get's the missing files you don't have in your local archive, and continues. 

For examples of other repos that exist: 

- [sec-edgar-downloader](https://github.com/jadchaar/sec-edgar-downloader)
- [python-edgar](https://github.com/edouardswiac/python-edgar/)

Furthermore, the hope of this package is to create parsers for repsective form types. A user could import the 10-K parser and call the Management Discussion and Analysis method to retrieve respective MD&A's from selected files. 

There are also plans to integrate directly with popular cloud providers given the scale of these filings. Processing 10-K/Q's alone requires TB's of storage.

##### Installation <a id='install' />
There are two primary methods of installing sec-utils. The first is via the python packaging index (pypi). The second is straight from source. 

To install from pypi:
```bash
pip install secutils
```

And to install from source:
```bash
git clone https://github.com/datawrestler/sec-utils && cd sec-utils
conda create --name sec_env python=3.7 pip
conda activate sec_env
pip install -r requirements.txt
pip install -e .
```

##### Usage <a id='usage' />
```bash
conda activate sec_env
python download_sec.py --output_dir=/mnt/sda/sec --form_types=S-1 --num_workers=-1 --start_year=2014 --end_year=2019 --quarters 1 2 3 4
```
Even more cleanly, you can coordinate long running jobs and keep track of your parameters by modifying this [example script](https://github.com/datawrestler/sec-utils/blob/master/examples/run.sh)

Make sure to make it executable on your system:
```bash
chmod +x run.sh
./run.sh
```

You can also generate a config file and use the config to control parameters of longer runs:
```python
from secutils.utils import generate_config
path_for_config = ''
generate_config(path_for_config)
```

then when calling the longer download run:
```bash
python -m secutils.download_sec --config_path='path_for_config'
```

A useful trick when working with remote servers is to direct output from a session to a file. Using screen also maintains a session even if you disconnect from ssh:
```bash
screen -dm -L python -m secutils.download_sec --config_path='path_for_config'
```

Additionally, users can leverage the API directly for more hands on work. An overview resides in an [example jupyter notebook](https://github.com/datawrestler/sec-utils/blob/master/examples/Getting%20Started.ipynb) with additional details below:
```python
from secutils.edgar import FormIDX
form = FormIDX(year=2017, quarter=1, seen_files=None, cache_dir=None, form_types=['10-K])
files = form.index_to_files()
form.master_index.head()
# CIK	Company Name	Form Type	Date Filed	Filename	fname
# 1000015	META GROUP INC	10-K	1998-03-31	edgar/data/1000015/0001000015-98-000009.txt	0001000015-98-000009.txt
# 1000112	CHEVY CHASE MASTER CREDIT CARD TRUST II	10-K	1998-03-27	edgar/data/1000112/0000920628-98-000038.txt	0000920628-98-000038.txt
# 1000179	PARAMOUNT FINANCIAL CORP	10-K	1998-03-30	edgar/data/1000179/0000950120-98-000108.txt	0000950120-98-000108.txt

# lets take a peek at attributes available to individual files:
ex = files[0]
msg = f"""
      Company Name: {ex.company_name}
      CIK Number: {ex.cik_number}
      Date Filed: {ex.date_filed}
      Form Type: {ex.form_type}
      File Name: {ex.file_name}
      Download URL: {ex.file_download_url}
      """
print(msg) 
# Company Name: OPTICAL CABLE CORP
# CIK Number: 1000230
# Date Filed: 2017-12-20 00:00:00
# Form Type: 10-K
# File Name: 0001437749-17-020936.txt
# Download URL: https://www.sec.gov/Archives/edgar/data/1000230/0001437749-17-020936.txt                                                                        
# get example file and download:
# to download our example file:
output_dir = '.'
ex.download_file(output_dir) # 200 is a successful download

# verify download 
import os
list(filter(lambda x: x.endswith('txt'), os.listdir(output_dir)))
# ['0001437749-17-020936.txt']
```

Getting hands on is great, however using the CLI does provide several advantages:
- Automatic directory structure creation
- Built in logging and caching
- Ability to resume training via download scanning
- Multi-threaded file downloading

##### Vision <a id='vision' />
The vision for this project extends far beyond it's current state of downloading index and SEC files from the Edgar database. Currently, parsing SEC files is tremendously difficult. There are numerous reasons for these difficulties including:

- No systematic tagging structure for SEC filings
- File submissions changed over the years
- Many different file types, header types, and content from one filing type to another

Given the above, parsing even a 10-K takes tremendous effort. The goal of this project is to bring together like minded individuals and take a stab at a systematic parsing effort with a consistent API. The future state of the project would allow users to download SEC filings and use convenient methods to retrieve particular sections of the filings. For instance, a user could do something like the following:

```python
from secutils.file_types import file_10k

file_path = '/path/to/10-K'
f = file_10k.from_path(file_path)

# and retrieve the management discussion and analysis section directly:
f.management_discussion()
# Here at XYZ company, we believe the following year will bring about great properity due to our R&D efforts in packages like secutils...
```
This would open up a world of opportunity for collaboration, text analytics research, and general business information gathering. 

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.0.3] - 2019-09-29
### Added
- Added ability to define run through yaml config object
- Added method `secutils.utils.generate_config` to generate sample config file
### Bug Fixes
- Removed pickling of SECContainer - logging is written to file
- Fixed progress bar updating when downloading files

[0.0.3]: https://github.com/datawrestler/sec-utils/releases/tag/v0.0.3

## [0.0.2] - 2019-09-22
### Added
- Minor refactoring of package import and documentation names

[0.0.2]: https://github.com/datawrestler/sec-utils/releases/tag/v0.0.2

## [0.0.1] - 2019-09-22
### Added
- Added Multi-threaded downloading
- Added Caching of index files
- Added Automatic directory structure buildout (i.e. downloading multiple file types w/dir structure: ftype --> year --> quarter --> files)
- Added Resume downloading
- Added Built in logging and download success tracker
- Added README, CHANGELOG, CONTRIBUTE

[0.0.1]: https://github.com/datawrestler/sec-utils/releases/tag/v0.0.1

secutils/edgar.py
import os
import io
import time
import random
import zipfile
import requests
import threading
import logging
import pickle as pkl
from pathlib import Path
from datetime import datetime
from urllib.error import HTTPError, URLError
from urllib.request import urlretrieve
from typing import List, Union, Optional
from urllib.parse import urlparse, urljoin

import ftfy
import pandas as pd
import validators
import httplib2

from secutils.utils import (
    _to_quarter, ValidateFields,
    _remove_bad_bytes, _check_cache_dir
)

logger = logging.getLogger(__name__)


class DocumentDownloaderThread(threading.Thread):
    """
    thread for downloading docs in parallel
    
    adapted from: https://github.com/freephys/Learning-Python-Design-Patterns/blob/master/2_singleton/crawler.py
    """
    def __init__(self, thread_id: int, name: str, output_dir: Path, cache_dir: Optional[str]=None) -> None:
        threading.Thread.__init__(self)
        self.name = name
        self.output_dir = output_dir
        self.cache_dir = cache_dir

    def run(self):
        download_docs(self.name, self.output_dir, self.cache_dir)


def download_docs(thread_name: str, output_dir: Path, cache_dir: Optional[str]=None) -> None:
    sec_container = SECContainer()
    # while we have pages where we have no downloaded docs
    while sec_container.to_visit:
        if '429' in sec_container.last_url_message:
            time.sleep(random.randint(1, 10))
        else:
            sec_file = sec_container.to_visit.pop()
            form_dir = build_dir_structure(output_dir, sec_file)
            time.sleep(random.randint(1, 10))
            urlmsg = sec_file.download_file(form_dir, cache_dir)
            if urlmsg == '200':
                sec_container.downloaded.add(sec_file)
                sec_container.pbar.update(1)
            else:
                setattr(sec_file, 'error_message')
                sec_container.download_error.add(sec_file)
            sec_container.last_url_message = urlmsg
            sec_container.pbar.set_postfix_str(f"Num success: {len(sec_container.downloaded)} -- Num errors: {len(sec_container.download_error)} -- Num remaining: {len(sec_container.to_visit)}")


class SECContainer(object):

    def __new__(cls):
        if not hasattr(cls, 'instance'):
            cls.instance = super(SECContainer, cls).__new__(cls)
        return cls.instance

class FileUtils(object):

    base_url = 'https://www.sec.gov/Archives/'

    def get_response(self, download_url: str) -> Union[str, str]:
        http = httplib2.Http()
        try:
            status, response = http.request(download_url)
        except Exception:
            raise RuntimeError(f'Unable to parse download url: {download_url}')
            
        try:
            response = response.decode('utf-8')
        except UnicodeDecodeError:
            response = ftfy.fix_text(response).decode('utf-8')
        return status, response

    def parse_url_to_parts(self, path: str) -> Union[str, str]:
        fname = path.split('/')[-1]
        # construct full url
        file_download_url = urljoin(self.base_url, path)
        assert validators.url(file_download_url), f"Invalid url: {file_download_url}"
        return fname, file_download_url


class File(FileUtils, ValidateFields):
    
    def __init__(self, form_type: str, company_name: str, cik_number: str, 
                date_filed: str, partial_url: str=None) -> None:

        # validate fields
        cik_number, company_name, form_type, date_filed, partial_url = self.validate_index_line(
            cik=cik_number, 
            company_name=company_name,
            form_type=form_type,
            date_filed=date_filed,
            partial_url=partial_url
        )
        self.form_type = form_type
        self.company_name = company_name
        self.cik_number = cik_number
        self.date_filed = date_filed
        self.year = self.date_filed.year
        self.quarter = _to_quarter(self.date_filed.month)
        self.file_name, self.file_download_url = self.parse_url_to_parts(partial_url)

    def to_row(self):
        return pd.DataFrame({
            'Form Type': self.form_type,
            'Company Name': self.company_name,
            'CIK': self.cik_number,
            'Date Filed': self.date_filed,
            'Year': self.year,
            'Quarter': self.quarter,
            'File Name': self.file_name,
            'Download URL': self.file_download_url,
            'Download File Path': getattr(self, 'download_file_dir', None)
        }, index=[0])

    def download_file(self, output_dir: str, cache_dir: Optional[str]=None) -> str:
        download_file_dir = os.path.join(output_dir, self.file_name)
        try:
            path, msg = urlretrieve(self.file_download_url, download_file_dir)
            # TODO do something better with these messages
            msg = '200'
            self.download_file_dir = download_file_dir
            if cache_dir:
                self.write_log_record(cache_dir)
        except (HTTPError, URLError) as e:
            msg = e
        return msg

    def write_log_record(self, cache_dir: str):
        parts = [self.cik_number, self.company_name, self.form_type, self.file_name, self.year, self.quarter, 
                self.file_download_url, self.download_file_dir]
        parts = list(map(str, parts))
        line = '|'.join(parts)
        with open(os.path.join(cache_dir, 'success.txt'), 'a') as outfile:
            outfile.write(line + '\n')
            outfile.close()



class FormIDX(object):
    """
    FormIDX is a utility class to capture master.idx zip files and construct a parsable data structure. 
    Each file user specifies to download is converted into a File class, where file downloads can occur.

    Parameters
    -------
    year: year of master.idx to download and parse
    quarter:  quarter of master.idx to download and parse
    seen_files: list of files already processed
    cache_dir: directory to cache master.idx files
    form_types: list of form types to download

    See Also:
    -------
    FileUtils
    ValidateFields

    Example:
    --------
    >>> from secutils.edgar import FormIDX
    >>> form = FormIDX(year=2017, quarter=1, seen_files=None, cache_dir=None, form_types=['10-K])
    >>> files = form.index_to_files()
    >>> form.master_index.head()
    >>> # CIK	Company Name	Form Type	Date Filed	Filename	fname
    >>> # 1000015	META GROUP INC	10-K	1998-03-31	edgar/data/1000015/0001000015-98-000009.txt	0001000015-98-000009.txt
    >>> # 1000112	CHEVY CHASE MASTER CREDIT CARD TRUST II	10-K	1998-03-27	edgar/data/1000112/0000920628-98-000038.txt	0000920628-98-000038.txt
    >>> # 1000179	PARAMOUNT FINANCIAL CORP	10-K	1998-03-30	edgar/data/1000179/0000950120-98-000108.txt	0000950120-98-000108.txt
    """
    full_index_url = 'https://www.sec.gov/Archives/edgar/full-index/{year}/QTR{quarter}/master.zip'

    def __init__(self, year: int, quarter: int, seen_files: Optional[List[str]] = None, 
                cache_dir: Optional[str]=None, form_types: Optional[List[str]]=None, 
                ciks: Optional[int]=None):
        self.year = year
        self.quarter = quarter
        self.download_url = self.full_index_url.format(year=year, quarter=quarter)
        self.seen_files = seen_files
        self.cache_dir = _check_cache_dir(cache_dir)
        self.ciks = ciks
        self.form_name = f"formidx-{self.year}-{self.quarter}.csv"
        self.form_types = form_types
        self.master_index = self._get_master_zip_index()

    def _get_master_zip_index(self) -> List[List[str]]:
        """download zip index files from Edgar db"""
        if self.cache_dir:
            cache_file = os.path.join(self.cache_dir, self.form_name)
        if self.cache_dir and os.path.exists(cache_file):
            master_index = pd.read_csv(cache_file)
        else:
            response = requests.get(self.download_url)
            status_code = response.status_code
            if status_code == 200:
                edgarzipfile = zipfile.ZipFile(io.BytesIO(response.content))
                edgarfile = edgarzipfile.open('master.idx').read()
                try:
                    edgarfile = edgarfile.decode('utf-8')
                    edgarfile = ftfy.fix_text(edgarfile)
                    lines = edgarfile.split('\n')
                except UnicodeDecodeError:
                    lines = edgarfile.split(b'\n')
                    lines = _remove_bad_bytes(lines)
                master_index = self._parse_index_lines(lines)
                if self.cache_dir:
                    master_index.to_csv(cache_file, index=False)
            else:
                logger.error(f"URL returned error ({status_code}): {self.year} - {self.quarter} - {self.download_url}")
                return None
        og_shape = master_index.shape[0]
        master_index = self._filter_form_type(master_index)
        master_index = self._filter_ciks(master_index)
        master_index = self._filter_seen_files(master_index)
        num_remaining_download = master_index.shape[0]
        msg = f"master index ({self.year}) - ({self.quarter}) - original shape: {og_shape} - remaining download: {num_remaining_download}"
        logger.info(msg)
        return master_index

    def _parse_index_lines(self, lines: List[str]) -> pd.DataFrame:
        split_line = lambda x: x.replace('\n', '').replace('\r', '').replace('\t', '').split('|')
        master_index = pd.DataFrame([split_line(line) for line in lines if line.count('|')==4])
        columns = ['CIK', 'Company Name', 'Form Type', 'Date Filed', 'Filename']
        master_index.columns = columns
        master_index = master_index.iloc[1:]
        master_index['fname'] = master_index['Filename'].apply(lambda x: x.split('/')[-1])
        return master_index

    def _filter_form_type(self, master_index: pd.DataFrame) -> pd.DataFrame:
        master_index['Form Type'] = master_index['Form Type'].apply(lambda x: x.strip())
        if self.form_types:
            unique_forms = master_index['Form Type'].unique().tolist()
            form_not_found = [form for form in self.form_types if form not in unique_forms]
            if len(form_not_found) > 0:
                msg = f"specified form type not found in master.idx ({self.year}) - ({self.quarter}) - form not found: {form_not_found}"
                logger.warning(msg)
            master_index = master_index.loc[master_index['Form Type'].isin(self.form_types)]
        return master_index

    def _filter_ciks(self, master_index: pd.DataFrame) -> pd.DataFrame:
        if self.ciks:
            self.ciks = [self.validate_cik(cik) for cik in self.ciks]
            master_index = master_index.loc[master_index['CIK'].isin(self.ciks)]
            msg = f"Found {master_index.shape[0]} files for CIK list"
            logger.info(msg)
        return master_index

    def _filter_seen_files(self, master_index: pd.DataFrame) -> pd.DataFrame:
        og_shape = master_index.shape[0]   
        if self.seen_files:
            master_index = master_index.loc[~master_index['fname'].isin(self.seen_files)]
            num_prior_download = og_shape-master_index.shape[0]
            msg = f"master index ({self.year}) - ({self.quarter}) - original shape: {og_shape} - num prior download: {num_prior_download}"
            logger.info(msg)
        return master_index

    def index_to_files(self) -> List[File]:
        files = []
        if isinstance(self.master_index, pd.DataFrame):
            for line in self.master_index.values:
                cik = line[0]
                company_name = line[1]
                form_type = line[2]
                date_filed = line[3]
                partial_url = line[4]

                files.append(File(
                    form_type=form_type,
                    company_name=company_name,
                    cik_number=cik,
                    date_filed=date_filed,
                    partial_url=partial_url,
                ))
        return files

def build_dir_structure(output_dir: str, sec_file: File) -> str: 
    # build output dir
    # modify form type due to forms like S-1/A
    form_type = sec_file.form_type.replace('/', '')
    output_dir = os.path.join(output_dir, form_type, str(sec_file.year), sec_file.quarter)
    if not os.path.exists(output_dir):
        os.makedirs(output_dir, exist_ok=True)
    return output_dir